In [2]:
import pandas as pd
import numpy as np
from google.colab import auth
auth.authenticate_user()
from googleapiclient.discovery import build
drive_service = build('drive', 'v3')
request = drive_service.files().get_media(fileId = '1XQgPGGrzOU24SGcujJOZ9mPpDDfjaV4o').execute()
import io
data = pd.read_csv(io.BytesIO(request))
data.head()

Year  Month  DayofMonth  ...  NASDelay  SecurityDelay  LateAircraftDelay
0  2008      1           3  ...       NaN            NaN                NaN
1  2008      1           3  ...       NaN            NaN                NaN
2  2008      1           3  ...       NaN            NaN                NaN
3  2008      1           3  ...       NaN            NaN                NaN
4  2008      1           3  ...       0.0            0.0               32.0

[5 rows x 29 columns]

In [3]:
data.shape

(7009728, 29)

**7009728** quan sát, 29 cột

In [4]:
data.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

In [5]:
data.describe() # data numerical

Year         Month  ...  SecurityDelay  LateAircraftDelay
count  7009728.0  7.009728e+06  ...   1.524735e+06       1.524735e+06
mean      2008.0  6.375130e+00  ...   7.497434e-02       2.077098e+01
std          0.0  3.406737e+00  ...   1.837940e+00       3.925964e+01
min       2008.0  1.000000e+00  ...   0.000000e+00       0.000000e+00
25%       2008.0  3.000000e+00  ...   0.000000e+00       0.000000e+00
50%       2008.0  6.000000e+00  ...   0.000000e+00       0.000000e+00
75%       2008.0  9.000000e+00  ...   0.000000e+00       2.600000e+01
max       2008.0  1.200000e+01  ...   3.920000e+02       1.316000e+03

[8 rows x 24 columns]

In [6]:
data.describe(include='O') # data category

UniqueCarrier  TailNum   Origin     Dest CancellationCode
count        7009728  6926363  7009728  7009728           137434
unique            20     5373      303      304                4
top               WN   N476HA      ATL      ATL                B
freq         1201754     4701   414513   414521            54904

Category data bao gồm: UniqueCarrier,	TailNum	Origin,	Dest,	CancellationCode, và Year,	Month,	DayofMonth,	DayOfWeek

In [7]:
missing_val_in_column = (data.isnull().sum())
print(missing_val_in_column[missing_val_in_column >0])

DepTime               136246
ArrTime               151649
TailNum                83365
ActualElapsedTime     154699
CRSElapsedTime           844
AirTime               154699
ArrDelay              154699
DepDelay              136246
TaxiIn                151649
TaxiOut               137058
CancellationCode     6872294
CarrierDelay         5484993
WeatherDelay         5484993
NASDelay             5484993
SecurityDelay        5484993
LateAircraftDelay    5484993
dtype: int64


CancellationCode     6872294 <br>
CarrierDelay         5484993<br>
WeatherDelay         5484993<br>
NASDelay             5484993<br>
SecurityDelay        5484993<br>
LateAircraftDelay  5484993<br>

Những features trên thiếu quá nhiều nên sẽ drop<br>

Feature: Year chỉ ra giấu cho năm 2018 nên drop.

### Handling missing values

In [0]:
data = data.drop(columns=['CancellationCode', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay', 'Year'])
data = data.dropna(subset=['DepTime', 'ArrTime', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
                          'DepDelay', 'TaxiIn', 'TaxiOut'])
# Tạo cột mới với encode cho target column ArrDelay
data['ArrDelay_ecd'] = np.where(data['ArrDelay'] >= 30, 1, 0)
# xóa cột ArrDelay đi vì ta đã có cột ArrDelay_ecd
data = data.drop(columns=['ArrDelay'])

In [9]:
# check missing value again
missing_val_in_column = (data.isnull().sum())
print(missing_val_in_column[missing_val_in_column >0])

Series([], dtype: int64)


Dữ liệu đã hết missing values

In [10]:
data.head()

Month  DayofMonth  DayOfWeek  ...  Cancelled  Diverted  ArrDelay_ecd
0      1           3          4  ...          0         0             0
1      1           3          4  ...          0         0             0
2      1           3          4  ...          0         0             0
3      1           3          4  ...          0         0             0
4      1           3          4  ...          0         0             1

[5 rows x 22 columns]

Target column giờ là ArrDelay_ecd

In [11]:
data['ArrDelay_ecd'].value_counts()

0    5927611
1     927413
Name: ArrDelay_ecd, dtype: int64

Số chuyến đến trễ **927413** so với đến sớm **5927611**<br>
Vậy dữ liệu imbalance

In [0]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn import preprocessing
from sklearn.preprocessing import OneHotEncoder

In [13]:
data.columns

Index(['Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime', 'ArrTime',
       'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'DepDelay', 'Origin',
       'Dest', 'Distance', 'TaxiIn', 'TaxiOut', 'Cancelled', 'Diverted',
       'ArrDelay_ecd'],
      dtype='object')

In [0]:
onehotencoder = OneHotEncoder()
data['UniqueCarrier'] = onehotencoder.fit_transform(np.array(data['UniqueCarrier']).reshape(-1,1)).toarray()

In [0]:
data = data.drop(columns=['TailNum', 'Origin', 'Dest'])

In [0]:
#data['Origin'] = onehotencoder.fit_transform(np.array(data['Origin']).reshape(-1,1)).toarray()
#data['Dest'] = onehotencoder.fit_transform(np.array(data['Dest']).reshape(-1,1)).toarray()

In [0]:
predictors = data.drop(['ArrDelay_ecd'], axis=1)
target = data["ArrDelay_ecd"]
x_train, x_val, y_train, y_val = train_test_split(predictors, target, test_size = 0.2, random_state = 0)

In [18]:
# Naive Bayes
from sklearn.naive_bayes import GaussianNB
gaussian = GaussianNB()
gaussian.fit(x_train, y_train)
y_pred = gaussian.predict(x_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.95      0.97   1185258
           1       0.74      0.89      0.81    185747

    accuracy                           0.94   1371005
   macro avg       0.86      0.92      0.89   1371005
weighted avg       0.95      0.94      0.95   1371005



In [19]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
y_pred = logreg.predict(x_val)
print(classification_report(y_val, y_pred))

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


              precision    recall  f1-score   support

           0       0.99      0.99      0.99   1185258
           1       0.94      0.95      0.95    185747

    accuracy                           0.99   1371005
   macro avg       0.97      0.97      0.97   1371005
weighted avg       0.99      0.99      0.99   1371005



In [20]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier
decisiontree = DecisionTreeClassifier()
decisiontree.fit(x_train, y_train)
y_pred = decisiontree.predict(x_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.99      1.00      1.00   1185258
           1       0.97      0.97      0.97    185747

    accuracy                           0.99   1371005
   macro avg       0.98      0.98      0.98   1371005
weighted avg       0.99      0.99      0.99   1371005



In [21]:
# Random Forest
from sklearn.ensemble import RandomForestClassifier

randomforest = RandomForestClassifier()
randomforest.fit(x_train, y_train)
y_pred = randomforest.predict(x_val)
print(classification_report(y_val, y_pred))

/usr/local/lib/python3.6/dist-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


              precision    recall  f1-score   support

           0       0.99      1.00      0.99   1185258
           1       0.97      0.92      0.95    185747

    accuracy                           0.99   1371005
   macro avg       0.98      0.96      0.97   1371005
weighted avg       0.99      0.99      0.99   1371005



In [22]:
# Gradient Boosting Classifier
from sklearn.ensemble import GradientBoostingClassifier

gbk = GradientBoostingClassifier()
gbk.fit(x_train, y_train)
y_pred = gbk.predict(x_val)
print(classification_report(y_val, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.99      0.98   1185258
           1       0.93      0.87      0.90    185747

    accuracy                           0.97   1371005
   macro avg       0.96      0.93      0.94   1371005
weighted avg       0.97      0.97      0.97   1371005



In [0]:
# Support Vector Machines
from sklearn.svm import SVC

svc = SVC()
svc.fit(x_train, y_train)
y_pred = svc.predict(x_val)
print(classification_report(y_val, y_pred))

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
